[examples] Add Remix example

mnajdova opened the pull request:
This PR adds example that uses Remix and TypeScript.

Fixes https://github.com/mui-org/material-ui/issues/29877

Same as https://github.com/mnajdova/remix-with-mui-typescript - the commits there can be reviewed, it was faster for testing in an isolated repository.

oliviertassinari reviewed (commented):
Should we add it in https://mui.com/getting-started/example-projects/#official-examples?

oliviertassinari commented on examples/remix-with-typescript/app/styles/dark.css:
```diff
@@ -0,0 +1,7 @@
+:root {
+  --color-foreground: hsl(0, 0%, 100%);
+  --color-background: hsl(0, 0%, 7%);
+  --color-links: hsl(213, 100%, 73%);
+  --color-links-hover: hsl(213, 100%, 80%);
+  --color-border: hsl(0, 0%, 25%);
+}
```
Why do we have this CSS file and the others? If the styles are really needed, I think that they should use the system.
```suggestion
```

mnajdova replied:
Let me remove these and add the same content as in the other templates. What do you think?

oliviertassinari commented on examples/remix-with-typescript/app/entry.server.tsx:
```diff
@@ -0,0 +1,60 @@
+        <meta name="viewport" content="width=device-width,initial-scale=1" />
+        ${css}
+        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap" />
+        <link href="https://fonts.googleapis.com/css2?family=Open+Sans:wght@300;400;600;700&display=swap" rel="stylesheet" />
+	      <link href="https://fonts.googleapis.com/css2?family=Poppins:wght@500&display=swap" rel="stylesheet" />
```
```suggestion
```

oliviertassinari commented on examples/remix-with-typescript/README.md:
````diff
@@ -0,0 +1,53 @@
+# Welcome to Remix!
+
+- [Remix Docs](https://remix.run/docs)
+
+## Development
+
+From your terminal:
+
+```sh
+npm run dev
+```
+
+This starts your app in development mode, rebuilding assets on file changes.
+
+## Deployment
+
+First, build your app for production:
+
+```sh
+npm run build
+```
+
+Then run the app in production mode:
+
+```sh
+npm start
+```
+
+Now you'll need to pick a host to deploy it to.
````
Should we follow the same template as the other examples? e.g. CRA

mnajdova commented:
> Should we add it in https://mui.com/getting-started/example-projects/#official-examples?

Yes, I will add it. I focused on making everything work in this first iteration.

oliviertassinari reviewed (commented):
Are we loading the font? I couldn't find something equivalent to 

https://github.com/mui-org/material-ui/blob/38ac830eedf60a2a40983d71cb4dd64e8f1abf4e/examples/nextjs/pages/_document.js#L14-L17

mnajdova commented on examples/remix-with-typescript/app/root.tsx:
```diff
@@ -0,0 +1,97 @@
+        <meta name="theme-color" content={theme.palette.primary.main} />
+        {title ? <title>{title}</title> : null}
+        <Meta />
+        <Links />
```
```suggestion
        <Links />
        <link
          rel="stylesheet"
          href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap"
        />
```

mnajdova commented:
> Are we loading the default font? I couldn't find something equivalent to

Good catch! I removed it when trying to scaffold everything from start. Added it back

siriwatknp reviewed (approved):
👍 cloned and tested!

siriwatknp commented on examples/nextjs-with-typescript/src/StylesContext.tsx:
```diff
@@ -0,0 +1,9 @@
+import { createContext } from 'react';
+
+export interface StyleContextData {
+  key: string;
+  ids: Array<string>;
+  css: string;
+}
+
+export default createContext<null | StyleContextData[]>(null);
```
Is this the correct place? it is in `nextjs-with-typescript`

mnajdova replied:
Ooopsy, I should remove it. Must have copied it in the wrong folder. Good catch!

krnbr commented:
Hi is there an example which supports SSR for MUI using REMIX + Express setup?

oliviertassinari commented on examples/remix-with-typescript/app/entry.server.tsx:
```diff
@@ -0,0 +1,57 @@
+  const emotionChunks = extractCriticalToChunks(html);
+
+  // Re-render including the extracted css.
+  const markup = renderToString(
```
I have just realized that we render twice. In https://remix.run/docs/en/v1/guides/styling

> It's unlikely this will effect performance in a significant way, React is pretty fast.

I don't buy it, not for one second. React is slow on the server side. At least, it's my personal experience at Onepixel, indexing about 10m pages with Next.js. Most of the time spent to get the first byte was on rendering the React tree (react apollo was half of this time). Looking back then, I can't imagine that it would have flied to render twice.

Could we look into the feasibility to render only once? :)

mnajdova replied:
Adding it to my Todo list

jurokapsiar replied:
@mnajdova I am wondering if this can be extended to render head+body in the first pass and only head in the second: https://gist.github.com/alexkuz/d91c9ce1380f4b528ebf262ad0e4f959

that would be so far the best way of using css-in-js with remix - until it gets better support for build time css-in-js.

mnajdova replied:
Hey @jurokapsiar :) That's a good idea. I plan to look into this next week when there is not much activity on the repo. I will open a PR with my findings.

mnajdova replied:
I had some time today, have created https://github.com/mui-org/material-ui/pull/30366 

I am basically inserting the style tags after an insertion point meta element that is added by the `root.tsx`. As far as I could test it works, but I may be missing something. It feels a bit hacky to parse the HTML string, but I don't have a better idea.

yehudamakarov commented:
where can i find this example in documentation? (assuming it is up somewhere already?)

yehudamakarov commented:
got it sorry: https://mui.com/getting-started/example-projects/#heading-official-examples
